File: backend/data/conn.go
package data

import (
	"context"
	"fmt"
	"log"
	"time"

	polygon "github.com/polygon-io/client-go/rest"
	polygonws "github.com/polygon-io/client-go/websocket"

	"github.com/jackc/pgx/v4/pgxpool"
)

type Conn struct {
	//Cache *redis.Client
	DB        *pgxpool.Pool
	Polygon   *polygon.Client
	PolygonWS *polygonws.Client
}

func InitConn(inContainer bool) (*Conn, func()) {
	//TODO change this sahit to use env vars as well
	var dbUrl string
	if inContainer {
		dbUrl = "postgres://postgres:pass@db:5432"
	} else {
		dbUrl = "postgres://postgres:pass@localhost:5432"
	}
	var dbConn *pgxpool.Pool
	var err error
	for true {
		dbConn, err = pgxpool.Connect(context.Background(), dbUrl)
		if err != nil {
			//if strings.Contains(err.Error(), "the database system is starting up") {
			log.Println("waiting for db")
			time.Sleep(5 * time.Second)
		} else {
			break
		}
	}
	/*
			   cache_url :="redis:6379"
			   cache := redis.NewClient(&redis.Options{Addr: cache_url,})
			   err = cache.Ping(context.Background()).Err()
			   if err != nil {
		           //god
			   }
	*/
	//return &Conn{DB: db, Cache: cache}
	//polygonConn := polygon.New(os.Getenv("POLYGON_API_KEY"))
	polygonConn := polygon.New("ogaqqkwU1pCi_x5fl97pGAyWtdhVLJYm")
	polygonWSConn, err := polygonws.New(polygonws.Config{
		APIKey: "ogaqqkwU1pCi_x5fl97pGAyWtdhVLJYm",
		Feed:   polygonws.RealTime,
		Market: polygonws.Stocks,
	})
	if err != nil {
		fmt.Printf("Error init polygonWs connection")
	}
	defer polygonWSConn.Close()
	if err := polygonWSConn.Connect(); err != nil {
		fmt.Printf("Error connecting to polygonWS")
	}
	conn := &Conn{DB: dbConn, Polygon: polygonConn, PolygonWS: polygonWSConn}

	cleanup := func() {
		conn.DB.Close()
	}
	return conn, cleanup
}

File: ml/conn.py
import time, psycopg2, redis

class Conn:

    def __init__(self, inside_container = True):
        if inside_container:
            cache_host = 'cache'
            db_host = 'db'
            tf_host = 'http://tf:8501/'
        else:
            cache_host = 'localhost'
            db_host = 'localhost'
            tf_host = 'http://localhost:8501/'
        while True:
            try:
                self.cache = redis.Redis(host=cache_host, port=6379)
                self.db = psycopg2.connect(host=db_host,port='5432',user='postgres',password='pass')
                self.tf = tf_host
                self.polygon = "ogaqqkwU1pCi_x5fl97pGAyWtdhVLJYm"
            except  psycopg2.OperationalError:
                print("waiting for db", flush = True)
                time.sleep(5)
            else:
                break


    def check_connection(self):
        try:
            self.cache.ping()
            self.db.ping()
        except:
            print('Connection error')
            self.__init__()
